### PR TITLE
DAOS-10358 control: Detect virtual net interfaces

### PR DESCRIPTION
**Pull request by kjacque**
The control plane can now detect virtual network interfaces as
part of the hardware topology. This allows end users to specify
a software-created interface like a sub-interface or a network
bridge for DAOS to use.

These changes also enable DAOS to detect multiple network devices
per fabric domain.

Signed-off-by: Kris Jacque <kristin.jacque@intel.com>

### CODE REVIEW

**Review by mjmac — APPROVED** [review]
LGTM

File: src/control/lib/hardware/pretty.go [hunk 1]
```diff
@@ -44,5 +44,16 @@ func PrintTopology(t *Topology, output io.Writer) error {
...
+	if len(t.VirtualDevices) > 0 {
+		fmt.Fprintln(ew, "Virtual Devices")
+
+		for _, dev := range t.VirtualDevices {
+			fmt.Fprintf(ew, "  %s (%s)\n", dev.Name, dev.Type)
+			if dev.BackingDevice != nil {
+				fmt.Fprintf(ew, "    backed by: %s\n", dev.BackingDevice)
+			}
+		}
+	}
```
[the remark this answers is not in the document]
Comment (kjacque) [reply src/control/lib/hardware/pretty.go:56]: I went back and forth on this myself, but I ended up reflecting the structure of the internal data, for simplicity. But yes, not all of them have a backing device (the loopback interface is a good example). In theory they could have more than one as well (for example, a bridge interface), but representing that got complicated fast, and I think it has limited value.

File: src/control/lib/hardware/topology.go [hunk 2]
```diff
@@ -182,6 +244,62 @@ func (t *Topology) Merge(newTopo *Topology) error {
...
+
+	for _, newVirt := range newTopo.VirtualDevices {
+		if curDev, found := curDevices[newVirt.Name]; found {
+			curVirt, ok := curDev.(*VirtualDevice)
+			if !ok {
+				return errors.Errorf("virtual device %q has same name as a hardware device", newVirt.Name)
+			}
+
+			if newVirt.Type != DeviceTypeUnknown {
+				curVirt.Type = newVirt.Type
```
Comment (knard-intel) [inline src/control/lib/hardware/topology.go:261]: Should we at least log a warning if curVirt.Type is different from DeviceTypeUnknown or is it OK to lost this information ?

File: src/control/lib/hardware/topology.go [hunk 3]
```diff
@@ -182,6 +244,62 @@ func (t *Topology) Merge(newTopo *Topology) error {
...
+			}
+
+			if err := mergeBackingDev(curDevices, curVirt, newVirt); err != nil {
```
Comment (knard-intel) [inline src/control/lib/hardware/topology.go:264]: Same comment as for the device type.

**Review by knard-intel — COMMENTED** [review]
Mostly OK for me, for what I understand.

File: src/control/cmd/daos_server/network.go [hunk 4]
```diff
@@ -69,18 +70,21 @@ func fabricInterfaceSetToHostFabric(fis *hardware.FabricInterfaceSet, filterProv
...
-		for _, provider := range fi.Providers.ToSlice() {
-			if filterProvider == "all" || strings.HasPrefix(provider, filterProvider) {
-				hf.AddInterface(&control.HostFabricInterface{
-					Provider:    provider,
-					Device:      name,
-					NumaNode:    uint32(fi.NUMANode),
-					NetDevClass: fi.DeviceClass,
-				})
+
+		for _, name := range netIFs.ToSlice() {
```
[the remark this answers is not in the document]
Comment (kjacque) [reply src/control/cmd/daos_server/network.go:78]: I vaguely recall that there was a reason why these two commands (daos_server network scan and daos_agent net-scan) are identical but not unified. We should probably do something like what was done with the dump topology command, which is kept in src/control/lib/hardware/hwprov. I'd like to deal with that in a separate PR though.

File: src/control/lib/hardware/topology.go (hunk 2 above)
Comment (kjacque) [reply src/control/lib/hardware/topology.go:261]: In theory I wouldn't really expect it to happen, though maybe that's a good reason to log it. I could address this in a follow-on PR.

File: src/control/lib/hardware/topology.go (hunk 3 above)
Comment (kjacque) [reply src/control/lib/hardware/topology.go:264]: We return the error, so I don't think it's necessary to log it here.

File: src/control/lib/hardware/topology.go (hunk 2 above)
Comment (knard-intel) [reply src/control/lib/hardware/topology.go:261]: OK

**Review by knard-intel — APPROVED** [review]
OK from my side.

File: src/control/lib/hardware/topology.go [hunk 5]
```diff
@@ -28,16 +28,59 @@ type (
...
+	return d.BackingDevice
+}
+
+// Topology is a hierarchy of hardware devices grouped under NUMA nodes.
+type Topology struct {
+	// NUMANodes is the set of NUMA nodes mapped by their ID.
+	NUMANodes NodeMap `json:"numa_nodes"`
+
+	// VirtualDevices is a set of virtual devices created in software that may have a
+	// hardware backing device.
```
Comment (tanabarr) [inline src/control/lib/hardware/topology.go:77]: nit: maybe advised that virtual devices are not pinned to a NUMA node and update the description of the topology struct type to indicate that virtual devices are not grouped under a numa node

**Review by tanabarr — APPROVED** [review]

**Pull request merged**